File: EleFourier/src/lib/Dft.cpp
/**
 * @copyright (C) 2012-2020 Euclid Science Ground Segment
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 3.0 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 *
 */

#include "EleFourier/Dft.h"

namespace Euclid {
namespace Fourier {

Fits::Position<2> RealForwardDftType::outShape(const Fits::Position<2>& shape) {
  return {shape[0] / 2 + 1, shape[1]};
}

Fits::Position<2> RealBackwardDftType::outShape(const Fits::Position<2>& shape) {
  return shape;
}

Fits::Position<2> ComplexForwardDftType::outShape(const Fits::Position<2>& shape) {
  return shape;
}

Fits::Position<2> ComplexBackwardDftType::outShape(const Fits::Position<2>& shape) {
  return shape;
}

fftw_plan RealForwardDftType::initFftwPlan(Fits::Raster<double, 3>& in, Fits::Raster<std::complex<double>, 3>& out) {
  const auto& shape = in.shape();
  const int width = static_cast<int>(shape[0]);
  const int height = static_cast<int>(shape[1]);
  int n[] = {height, width}; // FFTW ordering
  return fftw_plan_many_dft_r2c(
      2, // rank
      n, // n
      in.shape()[2], // howmany
      reinterpret_cast<double*>(in.data()), // in
      nullptr, // inembed
      1, // istride
      width * height, // idist
      reinterpret_cast<fftw_complex*>(out.data()), // out
      nullptr, // onembed
      1, // ostride
      (width / 2 + 1) * height, // odist
      FFTW_MEASURE); // FIXME other flags?
}

fftw_plan RealBackwardDftType::initFftwPlan(Fits::Raster<std::complex<double>, 3>& in, Fits::Raster<double, 3>& out) {
  const auto& shape = out.shape();
  const int width = static_cast<int>(shape[0]);
  const int height = static_cast<int>(shape[1]);
  int n[] = {height, width}; // FFTW ordering
  return fftw_plan_many_dft_c2r(
      2, // rank
      n, // n
      shape[2], // howmany
      reinterpret_cast<fftw_complex*>(in.data()), // in
      nullptr, // inembed
      1, // istride
      (width / 2 + 1) * height, // idist
      reinterpret_cast<double*>(out.data()), // out
      nullptr, // onembed
      1, // ostride
      width * height, // odist
      FFTW_MEASURE); // FIXME other flags?
}

fftw_plan ComplexForwardDftType::initFftwPlan(
    Fits::Raster<std::complex<double>, 3>& in,
    Fits::Raster<std::complex<double>, 3>& out) {
  const auto& shape = in.shape();
  const int width = static_cast<int>(shape[0]);
  const int height = static_cast<int>(shape[1]);
  int n[] = {height, width}; // FFTW ordering
  return fftw_plan_many_dft(
      2, // rank
      n,
      shape[2], // howmany,
      reinterpret_cast<fftw_complex*>(in.data()), // in
      nullptr, // inembed
      1, // istride
      width * height, // idist
      reinterpret_cast<fftw_complex*>(out.data()), // out
      nullptr, // onembed
      1, // ostride
      width * height, // odist
      FFTW_FORWARD, // sign
      FFTW_MEASURE); // FIXME other flags?
}

fftw_plan ComplexBackwardDftType::initFftwPlan(
    Fits::Raster<std::complex<double>, 3>& in,
    Fits::Raster<std::complex<double>, 3>& out) {
  const auto& shape = out.shape();
  const int width = static_cast<int>(shape[0]);
  const int height = static_cast<int>(shape[1]);
  int n[] = {height, width}; // FFTW ordering
  return fftw_plan_many_dft(
      2, // rank
      n,
      shape[2], // howmany,
      reinterpret_cast<fftw_complex*>(in.data()), // in
      nullptr, // inembed
      1, // istride
      width * height, // idist
      reinterpret_cast<fftw_complex*>(out.data()), // out
      nullptr, // onembed
      1, // ostride
      width * height, // odist
      FFTW_BACKWARD, // sign
      FFTW_MEASURE); // FIXME other flags?
}

} // namespace Fourier
} // namespace Euclid

File: PyFFTPlayground/python/PyFFTPlayground/DFT.py
#
# Copyright (C) 2012-2020 Euclid Science Ground Segment
#
# This library is free software; you can redistribute it and/or modify it under
# the terms of the GNU Lesser General Public License as published by the Free
# Software Foundation; either version 3.0 of the License, or (at your option)
# any later version.
#
# This library is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
# details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this library; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
#

from typing import Tuple, Any
import pyfftw


class DFT:
    """A very (very) light wrapper around pyfftw.FFTW.

    Keep it if at some point it is needed to wrap some methods in FFTW (to be able to use it with multiprocessing) or if
    more information related to the computation need to be stored.

    Attributes
    ----------
    - plan : pyfftw.FFTW
        DFT plan definition

    """

    def __init__(self, plan):
        self.plan = plan


def create_complex_plan(shape: Tuple[int], flags: Tuple[str]) -> "pyfftw.FFTW":
    """Create plan for monochromatic PSF computation. It is a complex inverse DFT
        Parameters
        ----------
        shape : tuple of ints
           shape of the input and output DFT

        flags : tuple of strings
           pyfftw planning flags
    """
    i = pyfftw.empty_aligned(shape, dtype="complex128")
    o = pyfftw.empty_aligned(shape, dtype="complex128")
    return DFT(pyfftw.FFTW(i, o, direction="FFTW_BACKWARD", flags=flags))


def create_real_plan(shape: Tuple[int], flags: Tuple[str]) -> "pyfftw.FFTW":
    """Create plan for Monochromatic PSF to MTF transform (real to complex DFT)
        Parameters
        ----------
        shape : tuple of ints
           shape of the input and output DFT

        flags : tuple of strings
           pyfftw planning flags
    """
    i = pyfftw.empty_aligned(shape, dtype="float64")
    output_shape = (i.shape[0], i.shape[-1] // 2 + 1)
    o = pyfftw.empty_aligned(output_shape, dtype="complex128")

    return DFT(pyfftw.FFTW(i, o, direction="FFTW_FORWARD", flags=flags))


def create_real_plan_backward(shape: Tuple[int], flags: Tuple[str]) -> "pyfftw.FFTW":
    """Create plan for broadband PSF (complex to real) - inverse
        Parameters
        ----------
        shape : tuple of ints
           shape of the input and output DFT

        flags : tuple of strings
           pyfftw planning flags
    """
    i = pyfftw.empty_aligned(shape, dtype="float64")
    output_shape = (i.shape[0], i.shape[-1] // 2 + 1)
    o = pyfftw.empty_aligned(output_shape, dtype="complex128")

    return DFT(pyfftw.FFTW(o, i, direction="FFTW_BACKWARD", flags=flags))
